File: Codigos estudiantes por lenguaje/C++/Julian_silva/algoritmos de ordenamientos/cycle sort.cpp
#include<bits/stdc++.h>
#include<cstdlib>
using namespace std;
void cycleSort(int arr[],int n){
    int writes=0;
    for(int cyclestart=0;cyclestart<n-2;cyclestart++){
        int item=arr[cyclestart];
        int pos=cyclestart;
        for(int i=cyclestart+1;i<n;i++){
            if(arr[i]<item){
                pos++;
            }
        }
        if(pos==cyclestart)continue;
        while(item==arr[pos]){
            pos++;
        }
        if(pos!=cyclestart){
            swap(item,arr[pos]);
            writes++;
        }
        while(pos!=cyclestart){
            pos=cyclestart;
            for(int i=cyclestart+1;i<n;i++){
                if(arr[i]<item){
                    pos++;
                }
            }
            while(item==arr[pos]){
                pos++;
            }
            if(item!=arr[pos]){
                swap(item,arr[pos]);
                writes++;
            }
        }
    }

}
int main() {
    int arr[]={1,8,3,9,10,2,4,10,45,-5};
    int n=sizeof(arr)/sizeof(arr[0]);
    cycleSort(arr,n);
    for(int i:arr){
        printf("%d ",i);
    }
}

File: Codigos estudiantes por lenguaje/PY/Bryann Valderrama/Algoritmos de Ordenamiento/cycleSort.py
'''Cycle Sort
 
   - Algoritmo in situ.
   - Su ventaja es que minimiza la utilizacion de memoria.
   - Mediante ciclos va ordenando los elementos.
   - Complejidad Tiempo:
        - Peor Caso: O(n^2)
        - Mejor Caso: O(n^2)
        - Caso Promedio: O(n^2)
'''

from sys import stdin, stdout
rl = stdin.readline
wr = stdout.write


def cycleSort(arr, n):

    writes = 0
    for cyclestart in range(n-2):
        item = arr[cyclestart]
        pos = cyclestart

        for i in range(cyclestart+1, n):
            if arr[i] < item:
                pos += 1

        if pos == cyclestart:
            continue

        while item == arr[pos]:
            pos += 1

        if pos != cyclestart:
            item, arr[pos] = arr[pos], item
            writes += 1

        while pos != cyclestart:
            pos = cyclestart
            for i in range(cyclestart+1, n):
                if arr[i] < item:
                    pos += 1
            while item == arr[pos]:
                pos += 1
            if item != arr[pos]:
                item, arr[pos] = arr[pos], item
                writes += 1


arr = list(map(int, rl().split()))
cycleSort(arr, len(arr))
print(*arr)
